src/Microsoft.ML.Dnn/ImageClassificationTransform.cs

```diff
@@ -720,6 +727,15 @@ private static IRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, Dat
                 (_evaluationStep, _) = AddEvaluationStep(_softMaxTensor, _labelTensor);
                 _softmaxTensorName = _softMaxTensor.name;
                 _predictionTensorName = _prediction.name;
+
+                VBuffer<ReadOnlyMemory<char>> keysVBuffer = default;
+                inputSchema[labelColumnName].GetKeyValues(ref keysVBuffer);
+                _keyValueAnnotations = keysVBuffer.DenseValues().ToArray();
+            }
+            else
```

Comment:
>else [](start = 12, length = 4)

I am assuming that this is meant for the case where you are loading the model from file. In that case, you need to use the key values that are deserialized from the model.

In general, I think the way the constructors for this class work is kind of confusing - I think there should be one constructor that is given all the information it needs, either from the deserialization `Create` method, or from the estimator `Fit` method. So everything inside `if (loadModel == false)` would be done in the estimator, as well as figuring out the class count. And the `Create` method for the `SignatureDataTransform` should not create a transformer directly, but instead create an estimator, then call `Fit` and `Transform`.

Reply:
Please open a separate issue for making two constructors, this suggestion is out of the scope of this change. 

---
In reply to: [328535147](https://github.com/dotnet/machinelearning/pull/4228#discussion_r328535147) [](ancestors = 328535147)